arch/x86/kvm/pkvm/mmu.c

```diff
@@ -207,6 +207,66 @@ static void set_host_mem_pgstate(unsigned long phys, unsigned long size,
 	}
 }
 
+static int check_host_mem_pgstate(unsigned long phys, unsigned long size,
+				  enum pkvm_page_state pgstate)
+{
+	unsigned long end = PAGE_ALIGN(phys + size);
+	struct pkvm_page *page;
+
+	if (!is_memory_range(phys, size))
+		return -EINVAL;
+
+	for (phys = PAGE_ALIGN_DOWN(phys); phys < end; phys += PAGE_SIZE) {
+		page = pkvm_phys_to_page(phys);
+		if (page->host_state != pgstate)
+			return -EPERM;
+	}
+
+	return 0;
+}
+
+struct page_ownership {
+	u64 *owner_id;
+};
+
+static int check_page_ownership_walker(struct pkvm_pgtable_visit_ctx *ctx,
+				       unsigned long walk_flags,
+				       void *const arg)
+{
+	struct pkvm_pgtable *pgt = ctx->pgt;
+	struct page_ownership *expt = arg;
+	void *ptep = ctx->ptep;
+
+	if (expt->owner_id) {
+		/* The pte should be a non-present entry to contain owner id. */
+		if (pgt->pgt_ops->pte_present(ptep))
+			return -EPERM;
+
+		if (*expt->owner_id != pkvm_pte_owner_id(pgt, ptep))
+			return -EPERM;
+	}
+
+	return 0;
+}
+
+static int check_page_ownership(struct pkvm_pgtable *pgt, unsigned long vaddr,
+				unsigned long size, u64 *expt_owner_id)
```

Comment:
Passing expt_owner_id and expt_pgstate by pointer makes the interface not the most convenient to use... So maybe it would be better to have separate functions e.g. check_page_owner() and check_page_owner_and_state(), to pass owner id and/or pgstate by value? (For example, they both could still use check_page_ownership_walker() as is.)